README.md
# Cookie Set plugin

A plugin that stores Google Analytics utm_parameter query string to a client cookie.

![Screenshot](resources/img/logo.png)

## Requirements

This plugin requires Craft CMS 3.0.0 or later.

## Installation

To install the plugin, follow these instructions.

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Then tell Composer to load the plugin:

        composer require webtrend/cookie-set

3. In the Control Panel, go to Settings → Plugins and click the “Install” button for Cookie Set.


## Using Cookie Set

		{% hook 'SetMyCookies' %}

Brought to you by **Webtrend**


CHANGELOG.md
# Cookie Set Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2018-12-11
## 1.0.1 - 2018-12-12
### Added
- Initial release


src/CookieSet.php
<?php
/**
 * Cookie Set plugin for Craft CMS 3.x
 *
 * A plugin that stores Google Analytics utm_parameter query string to a client cookies
 *
 * @link      https://google.com
 * @copyright Copyright (c) 2018 Webtrend
 */

namespace webtrend\cookieset;

use webtrend\cookieset\variables\CookieSetVariable;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\twig\variables\CraftVariable;

use yii\base\Event;

/**
 * Class CookieSet
 *
 * @author    Webtrend
 * @package   CookieSet
 * @since     1.0.1
 *
 */
class CookieSet extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * @var CookieSet
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $schemaVersion = '1.0.1';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('cookieSet', CookieSetVariable::class);
            }
        );

        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                }
            }
        );

        Craft::info(
            Craft::t(
                'cookie-set',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
        // Adds a hook 'SetMyCookies'
        // Syntax - {% hook 'SetMyCookies' %}
        Craft::$app->view->hook('SetMyCookies', function(&$context, $cookieName = 'tts', $expire = null, $path = '/', $domain = null, $secure = false, $httpOnly = false)
        {

            if (!$expire) {
                $expire = time() + 60*60*24*365*2;
            }
            isset($_GET['utm_term']) ? $utmParams['utm_term'] = $_GET['utm_term'] : NULL;
            isset($_GET['utm_campaign']) ? $utmParams['utm_campaign'] = $_GET['utm_campaign'] : NULL;
            isset($_GET['utm_source']) ? $utmParams['utm_source'] = $_GET['utm_source'] : NULL;
            isset($_GET['utm_medium']) ? $utmParams['utm_medium'] = $_GET['utm_medium'] : NULL;
            isset($_GET['utm_content']) ? $utmParams['utm_content'] = $_GET['utm_content'] : NULL;
            if (!isset($_COOKIE[$cookieName])) {
                $cookieData = array('firstContactDatetime' => date('c'));
                if (isset($_SERVER['HTTP_REFERER'])) {
                    $cookieData['referer'] = $_SERVER['HTTP_REFERER'];
                }
                else {
                    $cookieData['referer'] = 'direct';
                }

            } else  {
                $tempcookieData = (array) json_decode($_COOKIE[$cookieName]);
                if (isset($tempcookieData['firstContactDatetime']) ) {
                    $cookieData['firstContactDatetime'] = $tempcookieData['firstContactDatetime'];
                }
                if (isset($_SERVER['HTTP_REFERER'])) {
                    $cookieData['referer'] = $_SERVER['HTTP_REFERER'];
                }
                else {
                    if (isset($tempcookieData['referer']) ) {
                        $cookieData['referer'] = $tempcookieData['referer'];
                    }
                }
            }
            if (isset($utmParams) && is_array($utmParams)) {
                foreach ($utmParams as $key => $val) {
                    $cookieData[$key] = (isset($_REQUEST[$key])) ? $_REQUEST[$key] : null;
                }
                if (isset($cookieData)) {
                    setcookie($cookieName, json_encode($cookieData), $expire, $path, $domain, $secure, $httpOnly);
                }
            }
            //var_dump($cookieData);
        });
    }

    // Protected Methods
    // =========================================================================

}


src/variables/CookieSetVariable.php
<?php
/**
 * Cookie Set plugin for Craft CMS 3.x
 *
 * A plugin that stores Google Analytics utm_parameter query string to a client cookies
 *
 * @link      https://google.com
 * @copyright Copyright (c) 2018 Webtrend
 */

namespace webtrend\cookieset\variables;

use webtrend\cookieset\CookieSet;

use Craft;

/**
 * @author    Webtrend
 * @package   CookieSet
 * @since     1.0.1
 */
class CookieSetVariable
{
    /**
     * set() takes the same parameters as PHP's builtin setcookie();
     *
     * @param string $name
     * @param string $value
     * @param int $expire
     * @param string $path
     * @param string $domain
     * @param mixed $secure
     * @param mixed $httponly
     */
    function set($name = "", $value = "", $expire = null, $path = "/", $domain = "", $secure = false, $httponly = false)
    {
        if (!$expire) {
                $expire = time() + 60*60*24*365*2;
            }
        setcookie($name, $value, (int) $expire, $path, $domain, $secure, $httponly);
        $_COOKIE[$name] = $value;
    }
    /**
     * get() lets you retrieve the value of a cookie.
     *
     * @param mixed $name
     */
    function get($name, $querystring='',$referer='')
    {
        $cookie_string = '';
        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_campaign')
                {
                    $cookie_string .= '&utm_campaign='.urlencode($param);
                }
                if ($key =='utm_medium')
                {
                    $cookie_string .= '&utm_medium='.urlencode($param);
                }
                if ($key =='utm_term')
                {
                    $cookie_string .= '&utm_term='.urlencode($param);
                }
                if ($key =='utm_source')
                {
                    $cookie_string .= '&utm_source='.urlencode($param);
                }
                if ($key =='utm_content')
                {
                    $cookie_string .= '&utm_content='.urlencode($param);
                }
                
            }
            
        }
        if (empty($cookie_string))
            if (isset($_COOKIE[$name]))
            {
                $cookies = json_decode($_COOKIE[$name], true);
                if (json_last_error() == JSON_ERROR_NONE)
                {
                    $cookie_string = '';
                    foreach ($cookies as $key=>$cookie )
                    {
                        $cookie_string .= '&';
                        $cookie_string .= $key.'='.urlencode($cookie);
                    }
                    return $cookie_string;
                }
            }
            else
            {
                return '&utm_campaign=organic&referer='.urlencode($referer).'&firstContactDatetime='.date('c');
            }

        else {
            $cookie_string .= '&referer='.urlencode($referer).'&firstContactDatetime='.date('c');
            return $cookie_string;
        }
    }

    function getSource($name, $querystring='')
    {

        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_source')
                {
                    return $param;
                }

            }
            return '';

        }
        else if (isset($_COOKIE[$name]))
        {
            $cookies = json_decode($_COOKIE[$name], true);
            if (json_last_error() == JSON_ERROR_NONE)
            {
                foreach ($cookies as $key=>$cookie )
                {
                    if ($key =='utm_source')
                    {
                        return $cookie;
                    }

                }
                return '';
            }
        }
        else
        {
            return '';
        }


    }

    function getMedium($name, $querystring='')
    {

        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_medium')
                {
                    return $param;
                }

            }
            return '';

        }
        else if (isset($_COOKIE[$name]))
        {
            $cookies = json_decode($_COOKIE[$name], true);
            if (json_last_error() == JSON_ERROR_NONE)
            {
                foreach ($cookies as $key=>$cookie )
                {
                    if ($key =='utm_medium')
                    {
                        return $cookie;
                    }

                }
                return '';
            }
        }
        else
        {
            return '';
        }


    }
    function getCampaign($name, $querystring='')
    {

        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_campaign')
                {
                    return $param;
                }

            }
            return '';

        }
        else if (isset($_COOKIE[$name]))
        {
            $cookies = json_decode($_COOKIE[$name], true);
            if (json_last_error() == JSON_ERROR_NONE)
            {
                foreach ($cookies as $key=>$cookie )
                {
                    if ($key =='utm_campaign')
                    {
                        return $cookie;
                    }

                }
                return '';
            }
        }
        else
        {
            return '';
        }


    }
    function getTerm($name, $querystring='')
    {

        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_term')
                {
                    return $param;
                }

            }
            return '';

        }
        else if (isset($_COOKIE[$name]))
        {
            $cookies = json_decode($_COOKIE[$name], true);
            if (json_last_error() == JSON_ERROR_NONE)
            {
                foreach ($cookies as $key=>$cookie )
                {
                    if ($key =='utm_term')
                    {
                        return $cookie;
                    }

                }
                return '';
            }
        }
        else
        {
            return '';
        }


    }

    function getContent($name, $querystring='')
    {

        if (!empty($querystring))
        {
            parse_str($querystring, $params);
            foreach ($params as $key=>$param)
            {
                if ($key =='utm_content')
                {
                    return $param;
                }

            }
            return '';

        }
        else if (isset($_COOKIE[$name]))
        {
            $cookies = json_decode($_COOKIE[$name], true);
            if (json_last_error() == JSON_ERROR_NONE)
            {
                foreach ($cookies as $key=>$cookie )
                {
                    if ($key =='utm_content')
                    {
                        return $cookie;
                    }

                }
                return '';
            }
        }
        else
        {
            return '';
        }


    }
}


src/Cookieset.php
<?php
/**
 * Cookie Set plugin for Craft CMS 3.x
 *
 * A plugin that stores Google Analytics utm_parameter query string to a client cookie.
 *
 * @link      https://google.com
 * @copyright Copyright (c) 2018 Bhashkar Yadav
 */

namespace webtrend\cookieset;

use webtrend\cookieset\variables\CookieSetVariable;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\twig\variables\CraftVariable;

use yii\base\Event;

/**
 * Class CookieSet
 *
 * @author    Bhashkar Yadav
 * @package   CookieSet
 * @since     1.0.0
 *
 */
class CookieSet extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * @var CookieSet
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $schemaVersion = '1.0.0';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('cookieSet', CookieSetVariable::class);
            }
        );

        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                }
            }
        );

        Craft::info(
            Craft::t(
                'cookie-set',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
        // Adds a hook 'SetMyCookies'
        // Syntax - {% hook 'SetMyCookies' %}
        Craft::$app->view->hook('SetMyCookies', function(&$context, $cookieName = 'tts', $expire = null, $path = '/', $domain = null, $secure = false, $httpOnly = false)
        {

            if (!$expire) {
                $expire = time() + 60*60*24*365*2;
            }
            isset($_GET['utm_term']) ? $utmParams['utm_term'] = $_GET['utm_term'] : NULL;
            isset($_GET['utm_campaign']) ? $utmParams['utm_campaign'] = $_GET['utm_campaign'] : NULL;
            isset($_GET['utm_source']) ? $utmParams['utm_source'] = $_GET['utm_source'] : NULL;
            isset($_GET['utm_medium']) ? $utmParams['utm_medium'] = $_GET['utm_medium'] : NULL;
            isset($_GET['utm_content']) ? $utmParams['utm_content'] = $_GET['utm_content'] : NULL;
            if (!isset($_COOKIE[$cookieName])) {
                $cookieData = array('firstContactDatetime' => date('c'));
                if (isset($_SERVER['HTTP_REFERER'])) {
                    $cookieData['referer'] = $_SERVER['HTTP_REFERER'];
                }
                else {
                    $cookieData['referer'] = 'direct';
                }

            } else  {
                $tempcookieData = (array) json_decode($_COOKIE[$cookieName]);
                if (isset($tempcookieData['firstContactDatetime']) ) {
                    $cookieData['firstContactDatetime'] = $tempcookieData['firstContactDatetime'];
                }
                if (isset($_SERVER['HTTP_REFERER'])) {
                    $cookieData['referer'] = $_SERVER['HTTP_REFERER'];
                }
                else {
                    if (isset($tempcookieData['referer']) ) {
                        $cookieData['referer'] = $tempcookieData['referer'];
                    }
                }
            }
            if (isset($utmParams) && is_array($utmParams)) {
                foreach ($utmParams as $key => $val) {
                    $cookieData[$key] = (isset($_REQUEST[$key])) ? $_REQUEST[$key] : null;
                }
                if (isset($cookieData)) {
                    setcookie($cookieName, json_encode($cookieData), $expire, $path, $domain, $secure, $httpOnly);
                }
            }
            //var_dump($cookieData);
            
        });
    }

    // Protected Methods
    // =========================================================================

}
